### PR TITLE
[JENKINS-30101][JENKINS-30175] Simplify persistence design for temporarily offline status

### PR DESCRIPTION
**Pull request by Vlatombe**
There were complex code to synchronize transient state of `Computer` from `Node` and vice versa.
But since any temporarily offline cause is being set by a user and should be persisted in a `Node`, the `Computer` should actually delegate to `Node` instead of trying to synchronize its state.

Fixes [JENKINS-30101](https://issues.jenkins-ci.org/browse/JENKINS-30101) and [JENKINS-30175](https://issues.jenkins-ci.org/browse/JENKINS-30175)

### Proposed changelog entries

- Keep user offline reason when agent connects or disconnects for technical reasons.

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/hudson/model/Computer.java
Comment (Vlatombe) [inline core/src/main/java/hudson/model/Computer.java:707]: The field `offlineCause` is only for technical offline cause now. `temporarilyOfflineCause` only exists in `Node` now.

File: core/src/main/java/hudson/model/Computer.java
Comment (Vlatombe) [inline core/src/main/java/hudson/model/Computer.java:359]: Retaining previous behaviour : if a temporary offline cause is defined, from outside PoV, it replaces the technical offline cause.

File: core/src/main/java/hudson/model/Computer.java
Comment (jglick) [inline core/src/main/java/hudson/model/Computer.java:607]: https://github.com/jenkinsci/jenkins/blob/962a286bc67f2f3e65fe2842a1d7b12463d8cf89/core/src/main/java/jenkins/model/Jenkins.java#L2228 is `@CheckForNull` so this method should be too, and callers should in fact check for null.

IIRC it is legitimately possible for there to be a `SlaveComputer` with no corresponding `Slave`, at least in the case of permanent agents: the agent is still running an executor when the node is deleted. (Conversely, you can have a `Slave` with no `SlaveComputer`, if it has zero executors.) It is definitely a corner case and I do not think we need to expose any “temporarily offline” status in that case (the `SlaveComputer` will just be removed as soon as it is done with its current task), but it needs to handled and not throw an exception.

File: core/src/main/java/hudson/model/Computer.java
Comment (jglick) [inline core/src/main/java/hudson/model/Computer.java:701]:
```suggestion
    public void setTemporarilyOfflineCause(@CheckForNull OfflineCause temporarilyOfflineCause) {
```

File: core/src/main/java/hudson/model/Computer.java
Comment (jglick) [inline core/src/main/java/hudson/model/Computer.java:707]: Maybe clearer to move to `Node.setTemporaryOfflineCause`? (which IIUC will now only be called from this site)

File: core/src/main/java/hudson/model/Node.java
Comment (jglick) [inline core/src/main/java/hudson/model/Node.java:265]:
```suggestion
    boolean isTemporarilyOffline() {
```
?

**Review by jglick — CHANGES_REQUESTED** [review]
`getNodeOrDie` is probably wrong. Otherwise looks good.

File: core/src/main/java/hudson/model/Computer.java
Comment (jglick) [inline core/src/main/java/hudson/model/Computer.java:609]: I think this nullness check needs to be moved up into the callers, since some are just reading the status, not trying to set it, right?

File: core/src/main/java/hudson/model/Computer.java
Comment (jglick) [inline core/src/main/java/hudson/model/Computer.java:626]: I think you meant
```suggestion
        var node = getNode();
        return node != null && node.isTemporarilyOffline() || getChannel() == null;
```
or more simply
```suggestion
        return isTemporarilyOffline() || getChannel() == null;
```
If the node was deleted but the computer is still connected and busy, it should not be marked as offline.

**Review by jglick — COMMENTED** [review]
Just a nit in `isOffline`.

**Comment by mawinter69** [comment]
I think this will solve several issues that I also addressed in #6152.
What is missing here is showing the reason why an agent is temporarily offline on the agents page.

**Comment by mawinter69** [comment]
I think you will also need to adjust `setOfflineCause.jelly` in some way

**Comment by Vlatombe** [comment]
> What is missing here is showing the reason why an agent is temporarily offline on the agents page when that reason differs from the agent being disconnected (e.g. lost connection)

Looks like a separate change impacting UX.

**Review by timja — APPROVED** [review]
LGTM

**Review by jglick — APPROVED** [review]

**Review by mawinter69 — COMMENTED** [review]
Previously setting an agent temporarily offline also set the offline cause itself.
With this change once you set the agent temporarily offline it is not possible to get the offline cause that might have been set when disconnecting the agent after it was set temp offline.
Not sure if this will have implications for plugins.

`getOfflineCauseReason` will return an empty string now if an agent is only temp offline. This will break these tests I think:
https://github.com/jenkinsci/node-sharing-plugin/blob/ae54615d1213196f3ca91073caf7c2850758e85a/jth-tests/src/test/java/com/redhat/jenkins/nodesharing/SharedNodeCloudTest.java#L531
https://github.com/jenkinsci/openstack-cloud-plugin/blob/fc5384db4d5f1f4f2ebdec2aec7e28ce312cf0df/plugin/src/test/java/jenkins/plugins/openstack/compute/SingleUseSlaveTest.java#L137
Maybe  also tests in https://github.com/jenkinsci/java-client-api and others

**Comment by Vlatombe** [comment]
@mawinter69 Addressed your remark

**Review by jglick — APPROVED** [review]
OK I think. Seems worth running tests of the plugins mentioned.

File: core/src/main/java/hudson/model/Computer.java
Comment (mawinter69) [inline core/src/main/java/hudson/model/Computer.java:693]: The behaviour of this method changed. 
Previously if you set the `temporarilyOffline` to false, this ensured that the cause was ignored.
In the vsphere-cloud plugin they call this method with `false` and a cause.
https://github.com/jenkinsci/vsphere-cloud-plugin/blob/26db2ff0f928c0013bb1f448d65f933f61816efa/src/main/java/org/jenkinsci/plugins/vSphereCloudSlave.java#L321
Of course this is a bug in the plugin one can say

**Review by mawinter69 — CHANGES_REQUESTED** [review]

File: core/src/main/java/hudson/model/Computer.java
Comment (Vlatombe) [reply core/src/main/java/hudson/model/Computer.java:693]: Fixed

Comment (mawinter69) [inline core/src/main/java/hudson/model/Computer.java:693]: The javadoc is lost here especially that the second argument is only considered when the first is true. But maybe this doesn't matter as the method is now set to deprecated.

**Review by mawinter69 — APPROVED** [review]

File: core/src/main/java/hudson/model/Computer.java
Comment (Vlatombe) [reply core/src/main/java/hudson/model/Computer.java:693]: Yes, I'm dropping existing javadoc to avoid repeats, and only refer to the current methods.

**Comment by Vlatombe** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by mawinter69** [comment]
Can we mark this as fix for 
[JENKINS-30101](https://issues.jenkins-ci.org/browse/JENKINS-30101)
[JENKINS-30175](https://issues.jenkins-ci.org/browse/JENKINS-30175)
[JENKINS-50313](https://issues.jenkins-ci.org/browse/JENKINS-50313)

**Comment by Vlatombe** [comment]
Keeping [JENKINS-50313](https://issues.jenkins.io/browse/JENKINS-50313) out, since this PR doesn't cover showing both offline causes.

**Review by jglick — APPROVED** [review]

**Comment by MarkEWaite** [comment]
@vlatombe is there anything that we can do from Jenkins core to resolve the change in behavior that has been seen in the OpenStack cloud plugin or is a change needed in the plugin?

* https://github.com/jenkinsci/openstack-cloud-plugin/issues/384

**Comment by Vlatombe** [comment]
The `offlineCause` field semantic has changed slighly because the previous was ambiguous.

Filed https://github.com/jenkinsci/openstack-cloud-plugin/pull/385, compatibility is retained when using the getter.

**Pull request merged**
